### PR TITLE
SMS integration

### PR DESCRIPTION
**Pull request by Jeasmine**
SMS Integration

Emulates current channel email behavior, but in this case, for the new channel SMS.

Since most of the work between SMS and Email is the same, a parent class UserStateSecondaryChannelSynchronizer was created. This will hold all core functionality.

* Added setSMSNumber public function
* Add SMS observers
* Add SMS data to DeviceState
* Add UserStateSMSSynchronizer and UserStateSecondaryChannelSynchronizer files, share email and SMS logic under UserStateSecondaryChannelSynchronizer
* Move players and channel tests to SynchronizerIntegrationTests

Difference between SMS and email, not parent_player_id is sent (this is removed on logout PR), we relay on external_id

It is recommended to review commit by commit. The core implementation is done on the firsts commit, then on each test commit changes are made to make the feature truly work.

## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263)

### CODE REVIEW

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 1]
```diff
@@ -1960,9 +2090,7 @@ static void sendPurchases(JSONArray purchases, boolean newAsExisting, OneSignalR
             jsonBody.put("existing", true);
          jsonBody.put("purchases", purchases);
 
-         OneSignalRestClient.post("players/" + getUserId() + "/on_purchase", jsonBody, responseHandler);
-         if (getEmailId() != null)
-            OneSignalRestClient.post("players/" + getEmailId() + "/on_purchase", jsonBody, null);
+         OneSignalStateSynchronizer.sendPurchases(jsonBody, responseHandler);
       } catch (Throwable t) {
          Log(LOG_LEVEL.ERROR, "Failed to generate JSON for sendPurchases.", t);
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:2095]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java [hunk 2]
```diff
@@ -390,15 +417,15 @@ void onSuccess(String response) {
                         OneSignal.onesignalLog(OneSignal.LOG_LEVEL.DEBUG, "doCreateOrNewSession:response: " + response);
                         JSONObject jsonResponse = new JSONObject(response);
 
-                        if (jsonResponse.has("id")) {
-                            String newUserId = jsonResponse.optString("id");
+                        if (jsonResponse.has(ID)) {
+                            String newUserId = jsonResponse.optString(ID);
                             updateIdDependents(newUserId);
                             OneSignal.Log(OneSignal.LOG_LEVEL.INFO, "Device registered, UserId = " + newUserId);
```
Comment (aimiebell) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateSynchronizer.java:423]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/LocationIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+    @AfterClass
+    public static void afterEverything() throws Exception {
+        cleanUp();
+    }
+
+    @Test
+    @Config(shadows = {ShadowGoogleApiClientBuilder.class, ShadowGoogleApiClientCompatProxy.class, ShadowFusedLocationApiWrapper.class})
+    public void shouldUpdateAllLocationFieldsWhenTimeStampChanges() throws Exception {
+        ShadowApplication.getInstance().grantPermissions("android.permission.ACCESS_COARSE_LOCATION");
+        OneSignalInit();
```
Comment (aimiebell) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/LocationIntegrationTests.java:152]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/MainOneSignalClassRunner.java [hunk 4]
```diff
@@ -1377,7 +1342,6 @@ public void shouldUpdateNotificationTypesCorrectlyEvenWhenSetSubscriptionIsCalle
       assertEquals(1, ShadowOneSignalRestClient.lastPost.getInt("notification_types"));
    }
 
-
    @Test
    public void shouldAllowMultipleSetSubscription() throws Exception {
       OneSignalInit();
```
Comment (aimiebell) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/MainOneSignalClassRunner.java:1347]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,1831 @@
...
+    }
+
+    @AfterClass
+    public static void afterEverything() throws Exception {
+        cleanUp();
+    }
+
+    @Test
+    public void shouldSetEmail() throws Exception {
+        OneSignalInit();
```
Comment (aimiebell) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/SynchronizerIntegrationTests.java:191]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

**Review by jkasten2 — CHANGES_REQUESTED** [review]
Overall PR looks good, really like the work put into splitting up the tests here!

> Reviewed 10 of 22 files at r1, 14 of 15 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263#-:-MUhVb0v6c55L7igq1Wj:b-l4dw10)* status: all files reviewed, 2 unresolved discussions (waiting on @emawby, @gonzalonarbaiz, and @Jeasmine)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateEmailSynchronizer.java, line 48 at r3](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263#-MUhTOdU9PuWHGElokgm:-MUhTJCe1xFDrDsg0VIL:b-jrp6gh) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/f4ae88a007d0b78d053250def7d843fd576dd592/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateEmailSynchronizer.java#L48)):*
> ```Java
> 
>     @Override
>     void logoutSMS() {
> ```

This class has `logoutEmail()` and `logoutSMS()`, I would expect `logoutChannel()` instead. I haven't looked at your other PR so this might be addressed.
___
*[OneSignalSDK/unittest/src/test/java/com/test/onesignal/LocationIntegrationTests.java, line 86 at r3](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263#-MUhUkI_4TW8e3KqkkOa:-MUhUkI_4TW8e3KqkkOb:bl5s593) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/f4ae88a007d0b78d053250def7d843fd576dd592/OneSignalSDK/unittest/src/test/java/com/test/onesignal/LocationIntegrationTests.java#L86)):*
> ```Java
> )
> @RunWith(RobolectricTestRunner.class)
> public class LocationIntegrationTests {
> ```

:100: Nice clean up moving this to a different file!

**Review by Jeasmine — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263)* status: all files reviewed, 2 unresolved discussions (waiting on @emawby, @gonzalonarbaiz, and @jkasten2)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateEmailSynchronizer.java, line 48 at r3](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263#-MUhTOdU9PuWHGElokgm:-MUkFadN1C8HSxTjzVsf:bdtysui) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/f4ae88a007d0b78d053250def7d843fd576dd592/OneSignalSDK/onesignal/src/main/java/com/onesignal/UserStateEmailSynchronizer.java#L48)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

This class has `logoutEmail()` and `logoutSMS()`, I would expect `logoutChannel()` instead. I haven't looked at your other PR so this might be addressed.
</blockquote></details>

I Will address this on the logout PR! thanks for pointing this out, I agree 100%
___
*[OneSignalSDK/unittest/src/test/java/com/test/onesignal/LocationIntegrationTests.java, line 86 at r3](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263#-MUhUkI_4TW8e3KqkkOa:-MUkFhjUCsf8JodKz3sS:bdeavht) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/f4ae88a007d0b78d053250def7d843fd576dd592/OneSignalSDK/unittest/src/test/java/com/test/onesignal/LocationIntegrationTests.java#L86)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

:100: Nice clean up moving this to a different file!
</blockquote></details>

:fire:

**Review by jkasten2 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @emawby and @gonzalonarbaiz)

**Review by jkasten2 — APPROVED** [review]
> Reviewed 10 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1263)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @emawby and @gonzalonarbaiz)

**Review by emawby — APPROVED** [review]

**Pull request merged**